compiler/optimizer/LocalOpts.cpp

```diff
@@ -6811,84 +6819,65 @@ int32_t TR_InvariantArgumentPreexistence::perform()
             int32_t index = symbol->getParmSymbol()->getOrdinal();
             _parmInfo[index].setNotInvariant();
             --numInvariantArgs;
-            if (trace())
-               traceMsg(comp(), "PREX:    Arg %d (%s) is not invariant\n", index, node->getSymbolReference()->getName(comp()->getDebug()));
+            traceIfEnabled("PREX:    Arg %d (%s) is not invariant\n", index, node->getSymbolReference()->getName(comp()->getDebug()));
             }
          }
       }
 
-
+   TR::KnownObjectTable *knot = comp()->getKnownObjectTable();
    ListIterator<TR::ParameterSymbol> parms(&methodSymbol->getParameterList());
-   if (_isOutermostMethod)
-      {
-      if (trace())
-         traceMsg(comp(), "PREX:    Populating parmInfo of outermost method %s\n", feMethod->signature(trMemory()));
-
-      int32_t index = 0;
-      TR::ParameterSymbol *p;
-      p = parms.getFirst();
-      if (p && p->getOffset() == 0 && !feMethod->isStatic())
-         {
-         int32_t len = 0;
-         const char *sig = p->getTypeSignature(len);
-         TR_OpaqueClassBlock *clazz = fe()->getClassFromSignature(sig, len, feMethod);
-         _parmInfo[index].setSymbol(p);
-         if (clazz && !fe()->classHasBeenExtended(clazz))
-            {
-            _parmInfo[index].setClassIsCurrentlyFinal();
-            if (trace())
-               traceMsg(comp(), "PREX:      Receiver class is currently final\n");
-            }
-         _parmInfo[index].setClass(clazz);
-         if (trace())
-            traceMsg(comp(), "PREX:      Receiver class %p is %.*s\n", clazz, len, sig);
+   int32_t index = 0;
+   for (TR::ParameterSymbol *p = parms.getFirst(); p != NULL; p = parms.getNext(), index++)
+       {
+       ParmInfo &parmInfo = _parmInfo[index];
+       int32_t len = 0;
+       const char *sig = p->getTypeSignature(len);
+       if (*sig != 'L' || !parmInfo.isInvariant())
```

Comment:
This seems awfully Java specific - shouldn't this be hidden behind a Util API or something so different languages can supply their implementation here?

Reply:
Changed the code to check for the data type of the parm symbol.